core/src/test/scala/com/microsoft/azure/synapse/ml/causal/VerifySyntheticDiffInDiffEstimator.scala

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.causal
+
+import breeze.linalg.sum
+import breeze.stats.distributions.RandBasis
+import com.microsoft.azure.synapse.ml.core.test.fuzzing.{EstimatorFuzzing, TestObject}
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.ml.util.MLReadable
+import org.scalactic.{Equality, TolerantNumerics}
+import com.microsoft.azure.synapse.ml.causal.linalg._
+
+class VerifySyntheticDiffInDiffEstimator
+  extends EstimatorFuzzing[SyntheticDiffInDiffEstimator] {
+
+  import spark.implicits._
+
+  spark.sparkContext.setLogLevel("INFO")
+  Logger.getLogger("org").setLevel(Level.WARN)
+  Logger.getLogger("akka").setLevel(Level.WARN)
+
+  private lazy val rb = RandBasis.withSeed(47)
+  private implicit val equalityDouble: Equality[Double] = TolerantNumerics.tolerantDoubleEquality(1E-8)
+
+  private lazy val data = {
+    val rand1 = rb.gaussian(100, 1)
+    val controlPre1 = for {
+      unit <- 1 to 99
+      time <- 1 to 10
+    } yield (unit, time, 0, 0, rand1.sample + time)
+
+    val rand2 = rb.gaussian(50, 1)
+    val controlPre2 = for {
+      unit <- 100 to 100
+      time <- 1 to 10
+    } yield (unit, time, 0, 0, rand2.sample + time)
+
+    val rand3 = rb.gaussian(110, 1)
+    val controlPost1 = for {
+      unit <- 1 to 99
+      time <- 11 to 20
+    } yield (unit, time, 0, 1, rand3.sample)
+
+    val rand4 = rb.gaussian(60, 1)
+    val controlPost2 = for {
+      unit <- 100 to 100
+      time <- 11 to 20
+    } yield (unit, time, 0, 1, rand4.sample)
+
+    val rand5 = rb.gaussian(50, 1)
+    val treatPre = for {
+      unit <- 101 to 200
+      time <- 1 to 10
+    } yield (unit, time, 1, 0, rand5.sample)
+
+    val rand6 = rb.gaussian(40, 1)
+    val treatPost = for {
+      unit <- 101 to 200
+      time <- 11 to 20
+    } yield (unit, time, 1, 1, rand6.sample)
+
+    controlPre1 ++ controlPre2 ++ controlPost1 ++ controlPost2 ++ treatPre ++ treatPost
+  }
+
+  private lazy val df = data.toDF("Unit", "Time", "treatment", "postTreatment", "outcome")
+
+  private lazy val estimator = new SyntheticDiffInDiffEstimator()
+    .setTreatmentCol("treatment")
+    .setPostTreatmentCol("postTreatment")
+    .setOutcomeCol("outcome")
+    .setUnitCol("Unit")
+    .setTimeCol("Time")
+    .setMaxIter(500)
+  // Set LocalSolverThreshold to 1 to force Spark mode
+  // Spark mode and breeze mode should get same loss history and same solution
+  // .setLocalSolverThreshold(1)
+
+  test("SyntheticDiffInDiffEstimator can estimate the treatment effect") {
```

Comment:
might want to also test the codepath where it uses spark fitting ops

Reply:
I implemented the unit test for the linalg routine for both breeze mode and Spark mode, and since the MirrorDescent code is abstracted on top of linalg routines, I didn't implement the unit tests for the Spark fitting branch specifically. But I've just added a unit test for that with latest commit.